lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -6334,59 +7066,74 @@ bool ADContext::processDifferentiableAttribute(
     attr->setVJPName(vjpName);
   }
 
-  // If the JVP doesn't exist, need to synthesize it.
-  auto vjpGenerated = false;
-  if (!vjp) {
-    // Diagnose:
-    // - Functions with no return.
-    // - Functions with unsupported control flow.
-    if (diagnoseNoReturn(*this, original, invoker) ||
-        diagnoseUnsupportedControlFlow(*this, original, invoker))
-      return true;
-    
-    vjpGenerated = true;
-    vjp = createEmptyVJP(*this, original, attr, isAssocFnExported);
-    getGeneratedFunctions().push_back(vjp);
-    VJPEmitter emitter(*this, original, attr, vjp, invoker);
-    if (emitter.run()) {
-      return true;
-    }
-  }
-      
   // If the JVP doesn't exist, need to synthesize it.
   if (!jvp) {
     // Diagnose:
     // - Functions with no return.
     // - Functions with unsupported control flow.
-    if (vjpGenerated && (diagnoseNoReturn(*this, original, invoker) ||
+    if (RunJVPGeneration && (diagnoseNoReturn(*this, original, invoker) ||
         diagnoseUnsupportedControlFlow(*this, original, invoker)))
       return true;
 
     jvp = createEmptyJVP(*this, original, attr, isAssocFnExported);
     getGeneratedFunctions().push_back(jvp);
 
-    if (vjpGenerated) {
+    // For now, only run JVP emission if the flag is on and if there is no
+    // user defined VJP. If there is a user defined VJP but no JVP, that means
+    // the user should have provided a custom JVP as well since we likely
+    // cannot derive a custom JVP. Thus create empty body.
+    if (RunJVPGeneration && !vjp) {
       JVPEmitter emitter(*this, original, attr, jvp, invoker);
-      return emitter.run();
+      if (emitter.run())
+        return true;
     } else {
       LLVM_DEBUG(getADDebugStream()
                  << "Generating empty JVP for original @"
                  << original->getName() << '\n');
       // Create empty body of JVP if the user defined their own custom VJP.
-      // Return undef.
       auto *entry = jvp->createBasicBlock();
       createEntryArguments(jvp);
-      auto diffConv = jvp->getConventions();
       SILBuilder builder(entry);
       auto loc = jvp->getLocation();
-      builder.createReturn(loc, SILUndef::get(
-          jvp->mapTypeIntoContext(diffConv.getSILResultType()),
-          *jvp));
+      // Add a fatal error in case this function is called by the user.
+      auto fatalErrrorJvpType = SILFunctionType::get(
+          /*genericSig*/ nullptr,
+          SILFunctionType::ExtInfo().withRepresentation(
+              SILFunctionTypeRepresentation::Thin),
+          SILCoroutineKind::None, ParameterConvention::Direct_Unowned, {},
+          /*interfaceYields*/ {},
+          SILResultInfo(module.getASTContext().getNeverType(),
```

Comment:
Minor: please refactor the `SILResultInfo` into a local variable - maybe call it `neverResultInfo`.